src/ae_kqueue.c

```diff
@@ -101,6 +111,38 @@ static void aeApiFree(aeEventLoop *eventLoop) {
 
 static int aeApiAddEvent(aeEventLoop *eventLoop, int fd, int mask) {
     aeApiState *state = eventLoop->apidata;
+#ifdef USE_KQUEUE_BATCH
+    /* Instead of registering events to kqueue one by one, we buffer events and
+     * register them at once along with retrieving pending events in aeApiPoll. */
+    if (mask & AE_READABLE) {
+        EV_SET(state->changes + state->num_changes, fd, EVFILT_READ, EV_ADD, 0, 0, NULL);
+        /* The current changelist is full, register it to kqueue now and
+         * then rewind it to make room for follow-up events. */
+        if (++state->num_changes == MAX_QUEUED_EVENTS) {
+            if (kevent(state->kqfd, state->changes, state->num_changes, NULL, 0, NULL))
+                /* An error occurs while processing an element of the changelist,
+                 * this is unexpected and indicates somewhere went wrong.
+                 * We panic for this situation directly because we won't be unable to
+                 * learn about this failure later. */
+                panic("aeApiAddEvent: kevent, %s", strerror(errno));
+            state->num_changes = 0; /* rewind the changelist. */
+        }
+    }
+    if (mask & AE_WRITABLE) {
+        EV_SET(state->changes + state->num_changes, fd, EVFILT_WRITE, EV_ADD, 0, 0, NULL);
+        /* The current changelist is full, register it to kqueue now and
+         * then rewind it to make room for follow-up events. */
+        if (++state->num_changes == MAX_QUEUED_EVENTS) {
+            if (kevent(state->kqfd, state->changes, state->num_changes, NULL, 0, NULL))
+                /* An error occurs while processing an element of the changelist,
+                 * this is unexpected and indicates somewhere went wrong.
+                 * We panic for this situation directly because we won't be unable to
+                 * learn about this failure later. */
+                panic("aeApiAddEvent: kevent, %s", strerror(errno));
+            state->num_changes = 0; /* rewind the changelist. */
+        }
+    }
```

Comment:
this code block is almost identical to the block at L117-L130 with the only difference being the flag. single-source it?